### PR TITLE
Feature skills

### PR DESCRIPTION
**Pull request by samuelgrant**
## Management View
The management view allows leadership/senior FCs to create, update and delete skill sets. A skillset is a collection of skills that apply to a ship under a given name.
## TODO
- [ ] Backend task to fetch every pilot skill
  - Do hourly
- [ ] FC View (View all pilots)
  - Use the route `/commander/tools/skills?pilot={name}`
- [ ] Pilot View (Their Pilots **ONLY**) - This will probably wait until goon auth is implemented
  - Use the route `/my-skills`
- [x] Open up management access to Senior FCs
- [x] Get the state to update after a skill is added/edited through the modal.
- [ ] Comment code
- [ ] Clean up code

### CODE REVIEW

File: models/skills.js [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            }
+
+            cb(skills);
+        })
+    
+    }
+
+    module.updateSettings = (id, name, hulls, filter, isPublic, cb) => {
+        const hullsArray = hulls.split(',');
+        hulls = [];
```
Comment (mlohstroh) [inline models/skills.js]: looks like you're reusing a variable here, can we define another variable so we don't potentially stomp over data

File: models/skills.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            let hull = new Promise((resolve, reject) =>{
+                module.lookupID(hullsArray[i], (data) => {
+                    resolve(data);
+                });
+            })
+
+            hulls.push(hull);
+        }
+
+        Promise.all(hulls).then((hulls) => {
```
Comment (mlohstroh) [inline models/skills.js]: Same issue here about using the same `hull` variable.

I'd recommend something like `hullResults`

File: react/components/OpenMarket.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import React, { Component } from 'react';
+
+class OpenMarket extends Component {
+    constructor(props) {
```
Comment (mlohstroh) [inline react/components/OpenMarket.js]: constructor is no necessary if nothing else is in the body except `super(props)`

File: react/components/Panel.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        return this.props.bgclass;
+    }
+
+    render() {
+
+        let bgColor;
+        let borderColor;
+        let bgTextColor;
+        if(!!this.colorClass()) {
+            bgColor = ` bg-${this.colorClass()}`
```
Comment (mlohstroh) [inline react/components/Panel.js]: This interaction can be cleaned up with the `classNames` library I included.

```javascript
let titleClass = classNames("card-body", "bg-" + this.colorClass(), "bg-dark")

<div className={titleClass}>
```

File: react/components/Skills/Skill.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import React, { Component } from 'react';
+import OpenMarket from 'components/OpenMarket';
+
+
+class Skills extends Component {
+    constructor(props) {
```
Comment (mlohstroh) [inline react/components/Skills/Skill.js]: Unnecessary constructor

File: react/components/Skills/Skill.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    render() {
+        return(
+            <tr>
+                <td>{/* <i className="fas fa-check-circle text-success"></i> */}</td>
+                <td><OpenMarket item={this.props.skill} /></td>
+                <td>{this.props.skill.required}</td>
+                <td>{this.props.skill.recommended}</td>
+                <td><button className="btn btn-sm btn-primary" onClick={this.props.updateSkillHandler.bind(this, this.props.skill)}><i className="fas fa-edit"></i></button></td>
```
Comment (mlohstroh) [inline react/components/Skills/Skill.js]: You might have already tested this, but is the `.bind(this)` necessary in this context? There is nothing in this component that the context needs to care about unless I'm not seeing things correctly.

File: react/components/Skills/SkillCategory.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import React, { Component } from 'react';
+
+class skillCategory extends Component {
+    constructor(props) {
```
Comment (mlohstroh) [inline react/components/Skills/SkillCategory.js]: Unnecessary constructor

File: react/components/Skills/SkillsTable.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
+import React, { Component } from 'react';
+import Skills from 'components/Skills/Skill'
+
+class SkillsTable extends Component {
+    constructor(props) {
```
Comment (mlohstroh) [inline react/components/Skills/SkillsTable.js]: unnecessary constructor

File: react/skillsManagment.js [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+                        <div className="col-lg-4 col-md-6 col-sm-12">
+                            <SkillCategoryMenu skillList={this.state.skillsData.list} onSetSelected={this.updateSelectedSkillSet.bind(this)} />
+                        </div>
+                        <div className="col-lg-8 col-md-6 col-sm-12">
+                            <SkillSettings set={this.getSelectedSkillSet()} onChange={this.updateData.bind(this)} />
+                        </div>
+                    </div>
+                </section>
+
+                <Dialog title="Add a new Skill Set" id="newSkillSet">
```
Comment (mlohstroh) [inline react/skillsManagment.js]: Is this still valid with the other `SkillSettings` component? What's the difference here?

File: react/skillsManagment.js [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+                        <button className="btn btn-success"><i className="fas fa-check-circle"></i> Save</button>
+                    </form>
+                </Dialog>
+            </div>
+        );
+    }
+
+}
+
+console.log("Attaching to dom!");
```
Comment (mlohstroh) [inline react/skillsManagment.js]: we can probably drop this console log.

File: react/components/Panel.js (hunk 4 above)
Comment (samuelgrant) [reply react/components/Panel.js]: Will be patched shortly

File: react/components/Skills/Skill.js (hunk 6 above)
Comment (samuelgrant) [reply react/components/Skills/Skill.js]: Without the `bind(this, ` the application dies

File: react/components/Skills/UpdateSkillsDialog.js [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+            type: "post",
+            url: `/internal-api/v2/skills-managment/${this.props.setID._id}`,
+            data: {
+                name: this.state.skillName,
+                required: this.state.required,
+                recommended: this.state.recommended
+            }
+        }).done(() => {
+            $('#updateSkill').modal('hide');
+            // this.props.onChange(this);
```
Comment (mlohstroh) [inline react/components/Skills/UpdateSkillsDialog.js]: was this on purpose to be commented out? If it was on purpose, can we just remove it?

File: react/components/Skills/UpdateSkillsDialog.js [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+        }).fail((error) => {
+            console.log(error);
+        })
+    }
+    
+    render() {
+        let skillName;
+        if(!!this.getSkillName()) {
```
Comment (mlohstroh) [inline react/components/Skills/UpdateSkillsDialog.js]: Can't this just be `let skillName = this.getSkillName();`?

File: controllers/skillsController.js [hunk 13]
```diff
@@ -0,0 +1,116 @@
+var esi = require('eve-swagger');
+var setup = require('../setup.js');
+const users = require('../models/users.js')(setup);
+const skills = require('../models/skills.js')(setup);
+const log = require('../logger.js')(module);
+
+
+//Returns the Skills Managment page
+exports.managementIndex = (req, res) => {
+    if(!users.isRoleNumeric(req.user, 4)){
```
[the remark this answers is not in the document]
Comment (samuelgrant) [reply controllers/skillsController.js]: Not currently

File: models/skills.js (hunk 1 above)
Comment (samuelgrant) [reply models/skills.js]: We aren't stomping over data here. The hulls will be a string that gets put into the `hullsArray`. From there we have an array of objects that get populated further down the script.

**Pull request merged**